replay/models/ucb.py

```diff
@@ -0,0 +1,114 @@
+    can_predict_cold_items = True
+
+    def _fit(
+            self,
+            log: DataFrame,
+            user_features: Optional[DataFrame] = None,
+            item_features: Optional[DataFrame] = None,
+    ) -> None:
+        vals = log.select("relevance").where(
+            (sf.col("relevance") != 1) & (sf.col("relevance") != 0)
+        )
+        if vals.count() > 0:
+            raise ValueError("Relevance values in log must be 0 or 1")
+
+        items_counts = log.groupby("item_idx").agg(
+            sf.sum("relevance").alias("pos"),
+            sf.count("relevance").alias("total"),
+        )
+
+        full_count = log.count()
+        items_counts = items_counts.withColumn(
+            "relevance",
+            (sf.col("pos") / sf.col("total") + sf.sqrt(sf.log(sf.lit(2 * full_count)) / sf.col("total")))
+        )
+
+        self.item_popularity = items_counts.drop("pos", "total")
+        self.item_popularity.cache()
+
+    @property
+    def _dataframes(self):
+        return {"item_popularity": self.item_popularity}
+
+    def _clear_cache(self):
+        if hasattr(self, "item_popularity"):
+            self.item_popularity.unpersist()
+
+    # pylint: disable=too-many-arguments
+    def _predict(
+            self,
+            log: DataFrame,
+            k: int,
+            users: DataFrame,
+            items: DataFrame,
+            user_features: Optional[DataFrame] = None,
+            item_features: Optional[DataFrame] = None,
+            filter_seen_items: bool = True,
+    ) -> DataFrame:
+        selected_item_popularity = self.item_popularity.join(
+            items,
+            on="item_idx",
+            how="right",
+        ).fillna(value=1, subset=["relevance"]).withColumn(
+            "rank",
+            sf.row_number().over(Window.orderBy(sf.col("relevance").desc(),
+                                                sf.col("item_idx").desc())),
+        )
+
+        max_hist_len = 0
+        if filter_seen_items:
+            max_hist_len = (
+                (
+                    log.join(users, on="user_idx")
+                        .groupBy("user_idx")
+                        .agg(sf.countDistinct("item_idx").alias("items_count"))
+                )
+                    .select(sf.max("items_count"))
+                    .collect()[0][0]
+            )
+            # all users have empty history
+            if max_hist_len is None:
+                max_hist_len = 0
+
+        return users.crossJoin(
+            selected_item_popularity.filter(sf.col("rank") <= k + max_hist_len)
+        ).drop("rank")
+
+    def _predict_pairs(
+            self,
+            pairs: DataFrame,
+            log: Optional[DataFrame] = None,
+            user_features: Optional[DataFrame] = None,
+            item_features: Optional[DataFrame] = None,
+    ) -> DataFrame:
+        return pairs.join(self.item_popularity, on="item_idx", how="right")\
+            .fillna(value=1, subset=["relevance"])
```

Comment:
pls add your model to relevant tests from test_all_models

Reply:
ok

Reply:
wilson model does not in that test. This model is the same

Reply:
because it inherits predict with poprec, you model does not